File: dotnet/DesignPatterns/CreationalDesignPatterns/PrototypeDesignPattern/PrototypeDesignPattern.Example2/Models/Employee.cs
using PrototypeDesignPattern.Example2.Abstracts;

namespace PrototypeDesignPattern.Example2.Models;

public class Employee : IPrototype<Employee>
{
	public string Name { get; set; } = string.Empty;
	public string Position { get; set; } = string.Empty;
	public Address? Address { get; set; }

	public Employee ShallowCopy()
	{
		return (Employee)this.MemberwiseClone(); // built-in shallow copy method
	}

	public Employee DeepCopy()
	{
		return new Employee
		{
			Name = this.Name,
			Position = this.Position,
			Address = new Address
			{
				Street = this.Address?.Street ?? string.Empty,
				City = this.Address?.City ?? string.Empty
			}
		};
	}
}

File: nodejs/DesignPatterns/CreationalDesignPatterns/PrototypeDesignPattern/PrototypeDesignPattern.Example2/src/models/Employee.ts
import { IPrototype } from '../abstracts/IPrototype';
import { Address } from './Address';

export class Employee implements IPrototype<Employee> {
  name: string = '';
  position: string = '';
  address?: Address;

  shallowCopy(): Employee {
    return Object.assign(new Employee(), this);
  }

  deepCopy(): Employee {
    const copy = new Employee();
    copy.name = this.name;
    copy.position = this.position;
    copy.address = this.address
      ? { street: this.address.street, city: this.address.city }
      : undefined;
    return copy;
  }
}
